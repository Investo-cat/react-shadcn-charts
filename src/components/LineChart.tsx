import {
  LineChart as ReLineChart,
  Line,
  CartesianGrid,
  XAxis,
  Tooltip,
  ResponsiveContainer,
} from "recharts";
import type { CurveType } from "recharts/types/shape/Curve";

type Props = {
  data: { [key: string]: any }[];
  xKey: string;
  lineKey: string;
  color?: string;
  type?: CurveType;
  className?: string;
  isAnimationActive?: boolean;
};

export function LineChart({
  data,
  xKey,
  lineKey,
  color = "var(--color-line, #2563eb)",
  type = "natural",
  className,
  isAnimationActive = true,
}: Props) {
  return (
    <div className={className}>
      <ResponsiveContainer width="100%" height="100%">
        <ReLineChart
          data={data}
          margin={{
            left: 12,
            right: 12,
          }}
        >
          <CartesianGrid vertical={false} />
          <XAxis
            dataKey={xKey}
            tickLine={false}
            axisLine={false}
            tickMargin={8}
            interval="preserveStartEnd"
            tickFormatter={(value) =>
              typeof value === "string" ? value.slice(0, 3) : value
            }
          />
          <Tooltip />
          <Line
            dataKey={lineKey}
            type={type}
            stroke={color}
            strokeWidth={2}
            dot={false}
            isAnimationActive={isAnimationActive}
          />
        </ReLineChart>
      </ResponsiveContainer>
    </div>
  );
}
