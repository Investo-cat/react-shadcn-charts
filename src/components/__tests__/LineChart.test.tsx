import React from "react";

import { render, waitFor } from "@testing-library/react";
import { LineChart } from "../LineChart";

const mockData = [
  { month: "January", sales: 100 },
  { month: "February", sales: 200 },
  { month: "March", sales: 150 },
];

// Mock ResponsiveContainer to provide width/height
jest.mock("recharts", () => {
  const Original = jest.requireActual("recharts");

  return {
    ...Original,
    ResponsiveContainer: ({ children }: any) =>
      React.cloneElement(children, { width: 400, height: 300 }),
  };
});

describe("LineChart", () => {
  it("matches snapshot", () => {
    const { container } = render(
      <LineChart
        data={mockData}
        xKey="month"
        lineKey="sales"
        className="chart-class"
        isAnimationActive={false}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("applies custom className", () => {
    const { container } = render(
      <LineChart
        data={mockData}
        xKey="month"
        lineKey="sales"
        className="chart-class"
        isAnimationActive={false}
      />
    );

    expect(container.querySelector(".chart-class")).not.toBeNull();
  });

  it("renders SVG with line path", async () => {
    const { container } = render(
      <LineChart
        data={mockData}
        xKey="month"
        lineKey="sales"
        isAnimationActive={false}
      />
    );

    await waitFor(() => {
      const paths = container.querySelectorAll("path");
      expect(paths.length).toBeGreaterThan(0);
    });
  });

  it("renders XAxis ticks with formatted labels", async () => {
    const { findByText } = render(
      <LineChart
        data={mockData}
        xKey="month"
        lineKey="sales"
        isAnimationActive={false}
      />
    );

    expect(await findByText("Jan")).not.toBeNull();
    expect(await findByText("Feb")).not.toBeNull();
    expect(await findByText("Mar")).not.toBeNull();
  });
});
