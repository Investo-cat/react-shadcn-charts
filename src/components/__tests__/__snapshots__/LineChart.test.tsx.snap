// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LineChart matches snapshot 1`] = `
<div>
  <div
    class="chart-class"
  >
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 400px; height: 300px;"
    >
      <svg
        class="recharts-surface"
        height="300"
        style="width: 100%; height: 100%;"
        viewBox="0 0 400 300"
        width="400"
      >
        <title />
        <desc />
        <defs>
          <clippath
            id="recharts1-clip"
          >
            <rect
              height="270"
              width="376"
              x="12"
              y="0"
            />
          </clippath>
        </defs>
        <g
          class="recharts-cartesian-grid"
        >
          <g
            class="recharts-cartesian-grid-horizontal"
          >
            <line
              fill="none"
              height="270"
              stroke="#ccc"
              width="376"
              x="12"
              x1="12"
              x2="388"
              y="0"
              y1="270"
              y2="270"
            />
            <line
              fill="none"
              height="270"
              stroke="#ccc"
              width="376"
              x="12"
              x1="12"
              x2="388"
              y="0"
              y1="202.5"
              y2="202.5"
            />
            <line
              fill="none"
              height="270"
              stroke="#ccc"
              width="376"
              x="12"
              x1="12"
              x2="388"
              y="0"
              y1="135"
              y2="135"
            />
            <line
              fill="none"
              height="270"
              stroke="#ccc"
              width="376"
              x="12"
              x1="12"
              x2="388"
              y="0"
              y1="67.5"
              y2="67.5"
            />
            <line
              fill="none"
              height="270"
              stroke="#ccc"
              width="376"
              x="12"
              x1="12"
              x2="388"
              y="0"
              y1="0"
              y2="0"
            />
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="376"
                x="12"
                y="284"
              >
                <tspan
                  dy="0.71em"
                  x="12"
                >
                  Jan
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="376"
                x="200"
                y="284"
              >
                <tspan
                  dy="0.71em"
                  x="200"
                >
                  Feb
                </tspan>
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="376"
                x="388"
                y="284"
              >
                <tspan
                  dy="0.71em"
                  x="388"
                >
                  Mar
                </tspan>
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            d="M12,135C74.66666666666666,73.125,137.33333333333331,11.25,200,0C262.6666666666667,-11.25,325.33333333333337,28.125,388,67.5"
            fill="none"
            height="270"
            stroke="var(--color-line, #2563eb)"
            stroke-width="2"
            width="376"
          />
          <g
            class="recharts-layer"
          />
        </g>
      </svg>
      <div
        class="recharts-tooltip-wrapper"
        style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"
        tabindex="-1"
      >
        <div
          class="recharts-default-tooltip"
          style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid rgb(204, 204, 204); white-space: nowrap;"
        >
          <p
            class="recharts-tooltip-label"
            style="margin: 0px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
